### PR TITLE
Upgrade RKE2 clusters

### PR DESCRIPTION
**Pull request by atanasdinov**
- Introduces a `manifests` directory storing the manifests for the different release versions
- Introduces RKE2 upgrade flow

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -91,7 +91,7 @@ run: manifests generate fmt vet ## Run a controller from your host.
 # More info: https://docs.docker.com/develop/develop-images/build_enhancements/
 .PHONY: docker-build
 docker-build: ## Build docker image with the manager.
-	$(CONTAINER_TOOL) build -t ${IMG} .
+	$(CONTAINER_TOOL) build -t ${IMG} --arch linux/amd64 .
```
Comment (atanasdinov) [inline Makefile:94]: [rke2-upgrade](https://hub.docker.com/r/rancher/rke2-upgrade) images are only available for amd64, so this allows for easier testing regardless of the host used to build the controller.

File: api/v1alpha1/upgradeplan_types.go [hunk 2]
```diff
@@ -20,6 +20,22 @@ import (
...
+	UpgradePending = "Pending"
+
+	// UpgradeInProgress indicates that the upgrade process has started (e.g. by creating plans.upgrade.cattle.io resources).
+	UpgradeInProgress = "InProgress"
+
+	// UpgradeSucceeded indicates that the upgrade process has been successful.
+	UpgradeSucceeded = "Upgraded"
+
+	// UpgradeFailed indicates that an error occurred during the upgrade process.
+	UpgradeFailed = "Failed"
```
Comment (atanasdinov) [inline api/v1alpha1/upgradeplan_types.go:36]: Not yet used, but I felt like including it.

File: internal/controller/upgradeplan_controller.go [hunk 3]
```diff
@@ -18,47 +18,223 @@ package controller
...
+}
+
+func (r *UpgradePlanReconciler) recordCreatedPlan(upgradePlan *lifecyclev1alpha1.UpgradePlan, name, namespace string) {
+	r.Recorder.Eventf(upgradePlan, corev1.EventTypeNormal, "PlanCreated", "Upgrade plan created: %s/%s", namespace, name)
+}
+
+func (r *UpgradePlanReconciler) reconcileKubernetes(ctx context.Context, upgradePlan *lifecyclev1alpha1.UpgradePlan, kubernetesVersion string) (ctrl.Result, error) {
+	var requeue bool
+
+	controlPlanePlan := &upgradecattlev1.Plan{}
```
Comment (atanasdinov) [inline internal/controller/upgradeplan_controller.go:124]: The flow of checking whether the plans exist and creating them if not, is similar enough but I opted for not unifying it just yet.

File: internal/plan/plan.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+import (
+	upgradecattlev1 "github.com/rancher/system-upgrade-controller/pkg/apis/upgrade.cattle.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Option func(plan *upgradecattlev1.Plan)
+
+func New(name string, opts ...Option) *upgradecattlev1.Plan {
```
Comment (atanasdinov) [inline internal/plan/plan.go:11]: I added this as a convenient way of creating the different plans - let me know what you think.

**Comment by atanasdinov** [comment]
Note that this only supports clusters where both control plane and worker nodes are present. The other cases will be covered in a follow-up.

**Review by ipetrov117 — COMMENTED** [review]
Left out some comments, but IMO the direction of the reconciliation flow is correct. 

With the introduction of OS and Helm component upgrades some of the functionality that is currently under `kubernetes.go` would need to be moved IMO, but that will be handled later down the line. 

Once we get to a "stable" PoC for the controller we should look back at what we are using and what is obsolete from the generated controller skeleton and do some cleanup. Great work on this!

File: api/v1alpha1/upgradeplan_types.go [hunk 5]
```diff
@@ -20,6 +20,22 @@ import (
...
+)
+
 // UpgradePlanSpec defines the desired state of UpgradePlan
 type UpgradePlanSpec struct {
```
[the remark this answers is not in the document]
Comment (atanasdinov) [reply api/v1alpha1/upgradeplan_types.go:40]: Yes, and I've put this on my notes but validation is one of the things I'd rather add later than try to fight now.

[the remark this answers is not in the document]
Comment (ipetrov117) [reply api/v1alpha1/upgradeplan_types.go:40]: Fine by me, as long as we have it in our todos.

File: api/v1alpha1/upgradeplan_types.go [hunk 6]
```diff
@@ -29,8 +45,12 @@ type UpgradePlanSpec struct {
 
 // UpgradePlanStatus defines the observed state of UpgradePlan
 type UpgradePlanStatus struct {
-	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// +listType=map
+	// +listMapKey=type
+	// +patchStrategy=merge
```
[the remark this answers is not in the document]
Comment (atanasdinov) [reply api/v1alpha1/upgradeplan_types.go:50]: I haven't tested the integration between the Makefile comments. Just copy pasted from the [official Kubernetes sources](https://github.com/kubernetes/apimachinery/blob/release-1.28/pkg/apis/meta/v1/types.go#L1482-L1542).

File: internal/controller/upgradeplan_controller.go [hunk 7]
```diff
@@ -18,47 +18,223 @@ package controller
...
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+
+	Releases map[string]*release.Release
 }
 
 // +kubebuilder:rbac:groups=lifecycle.suse.com,resources=upgradeplans,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=lifecycle.suse.com,resources=upgradeplans/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=lifecycle.suse.com,resources=upgradeplans/finalizers,verbs=update
+// +kubebuilder:rbac:groups=upgrade.cattle.io,resources=plans,verbs=create;list;get;watch
```
[the remark this answers is not in the document]
Comment (atanasdinov) [reply internal/controller/upgradeplan_controller.go:55]: Well, this is actually the source of truth. We adjust these comments and then `make` and / or `make manifests` make the modifications to the files under `config/`.

File: internal/controller/upgradeplan_controller.go [hunk 8]
```diff
@@ -18,47 +18,223 @@ package controller
...
 
+	if err := r.Get(ctx, req.NamespacedName, &plan); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	logger.Info("Upgrade to the SUSE platform requested",
-		"version", upgradePlan.Spec.ReleaseVersion)
+	logger.Info("Reconciling UpgradePlan")
+
+	planCopy := plan.DeepCopy()
```
[the remark this answers is not in the document]
Comment (atanasdinov) [reply internal/controller/upgradeplan_controller.go:72]: As previously discussed, this was added in order to reduce the calls to `Status.Update()` to a minimum, but I'll try to remove it.

File: internal/upgrade/kubernetes.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+			controlPlanePlanName,
+		},
+		Image: rke2UpgradeImage,
+	}
+	workerPlan.Spec.Upgrade = &upgradecattlev1.ContainerSpec{
+		Image: rke2UpgradeImage,
+	}
+	workerPlan.Spec.Version = version
+	workerPlan.Spec.Cordon = true
+	workerPlan.Spec.Drain = &upgradecattlev1.DrainSpec{
```
[the remark this answers is not in the document]
Comment (atanasdinov) [reply internal/upgrade/kubernetes.go:113]: Agree that we should probably revisit this. I'd stick to the most simple flow, as described in our Day 2 docs, and only start adjusting once the controller is in a good state to start actively testing.

**Review by ipetrov117 — APPROVED** [review]
Code looks good IMO. Once we start adding other Day 2 components we will have a better idea whether we should rework the reconcile logic (don't think we would need to), or reorder existing code. Great work!

File: config/rbac/role.yaml [hunk 10]
```diff
@@ -4,6 +4,20 @@ kind: ClusterRole
...
+  - ""
+  resources:
+  - nodes
+  verbs:
+  - list
+  - watch
+- apiGroups:
+  - ""
+  resources:
+  - events
```
Comment (Kristian-ZH) [inline config/rbac/role.yaml:17]: Why it needs permissions for events permissions?

**Review by Kristian-ZH — APPROVED** [review]
/lgtm

File: config/rbac/role.yaml (hunk 10 above)
Comment (atanasdinov) [reply config/rbac/role.yaml:17]: It's more of an experimentation at this point but I included [event publishing](https://github.com/suse-edge/upgrade-controller/pull/4/files#diff-21c33495c902058032ff4e15139880e160e4ad997a2db042dbac55bb5c3dc426R103) for created plans.

**Pull request merged**